internal/kafka/franz_client_test.go

```diff
@@ -0,0 +1,337 @@
+			return err
+		}
+		defer client.Close()
+		return client.Ping(context.Background())
+	}
+
+	t.Run("tls_valid_ca", func(t *testing.T) {
+		t.Parallel()
+		tlsConfig := configtls.NewDefaultClientConfig()
+		tlsConfig.CAPem = configopaque.String(
+			pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: caCert.Raw}),
+		)
+		assert.NoError(t, tryConnect(tlsConfig))
+	})
+
+	t.Run("tls_insecure_skip_verify", func(t *testing.T) {
+		t.Parallel()
+		tlsConfig := configtls.NewDefaultClientConfig()
+		tlsConfig.InsecureSkipVerify = true
+		require.NoError(t, tryConnect(tlsConfig))
+	})
+
+	t.Run("tls_unknown_ca", func(t *testing.T) {
+		t.Parallel()
+		config := configtls.NewDefaultClientConfig()
+		err := tryConnect(config)
+		require.Error(t, err)
+		assert.ErrorContains(t, err, "x509: certificate signed by unknown authority")
+	})
+
+	t.Run("plaintext", func(t *testing.T) {
+		t.Parallel()
+		// Should fail because the server expects TLS.
+		require.Error(t, tryConnect(configtls.ClientConfig{}))
+	})
+}
+
+func TestNewFranzGoSyncProducerCompression(t *testing.T) {
+	compressionAlgos := []string{"none", "gzip", "snappy", "lz4", "zstd"}
+	for i, compressionAlgo := range compressionAlgos {
+		t.Run(compressionAlgo, func(t *testing.T) {
+			t.Parallel()
+			validTopic := fmt.Sprintf("test-topic-%d", i)
+			// Create a cluster with a superuser that uses PLAIN authentication.
+			cluster, clientConfig := kafkatest.NewCluster(t, kfake.EnableSASL(),
+				kfake.Superuser(PLAIN, "plain_user", "plain_password"),
+				kfake.SeedTopics(1, validTopic),
+			)
+			clientConfig.Authentication.SASL = &configkafka.SASLConfig{
+				Mechanism: PLAIN,
+				Username:  "plain_user",
+				Password:  "plain_password",
+				Version:   1, // kfake only supports version 1
+			}
+			prodCfg := configkafka.NewDefaultProducerConfig()
+			prodCfg.Compression = compressionAlgo
+
+			tl := zaptest.NewLogger(t, zaptest.Level(zap.WarnLevel))
+			client, err := NewFranzSyncProducer(clientConfig, prodCfg, time.Second, tl)
+			require.NoError(t, err)
+			defer client.Close()
+
+			ctx, cancel := context.WithTimeoutCause(context.Background(), time.Second,
+				errors.New("Failed to connect to Kafka cluster"),
+			)
+			defer cancel()
+			require.NoError(t, client.Ping(ctx))
+
+			body := "test message"
+			cluster.ControlKey(int16(kmsg.Produce), func(kreq kmsg.Request) (kmsg.Response, error, bool) {
+				preq := kreq.(*kmsg.ProduceRequest)
+				assert.Len(t, preq.Topics, 1, "expected one topic in produce request")
+				assert.Equal(t, validTopic, preq.Topics[0].Topic, "produced to wrong topic")
+				assert.Len(t, preq.Topics[0].Partitions, 1, "expected one partition in produce request")
+				assert.Equal(t, int32(0), preq.Topics[0].Partitions[0].Partition, "produced to wrong partition")
+
+				var rb kmsg.RecordBatch
+				require.NoError(t, rb.ReadFrom(preq.Topics[0].Partitions[0].Records))
+				assert.Equal(t, int32(1), rb.NumRecords, "expected one record in produce request")
```

Comment:
Should we check the `rb.Attributes` field, to confirm the data is compressed with the expected codec?